File: android/app/src/main/java/com/alexivanov/snapsell/ui/inventory/InventoryScreen.kt
package com.alexivanov.snapsell.ui.inventory

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.RectangleShape
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import com.alexivanov.snapsell.AppContainer
import com.alexivanov.snapsell.R
import com.alexivanov.snapsell.data.local.ListingWithItems
import com.alexivanov.snapsell.domain.ListingStatus
import com.alexivanov.snapsell.domain.Money
import com.alexivanov.snapsell.ui.common.CutoutTile
import com.alexivanov.snapsell.ui.common.GhostButton
import com.alexivanov.snapsell.ui.common.IconBox
import com.alexivanov.snapsell.ui.common.LucideIcon
import com.alexivanov.snapsell.ui.common.MicroLabel
import com.alexivanov.snapsell.ui.common.Rule
import com.alexivanov.snapsell.ui.common.SnapTopBar
import com.alexivanov.snapsell.ui.common.StatusChip
import com.alexivanov.snapsell.ui.theme.SnapType
import com.alexivanov.snapsell.ui.theme.snapColors
import java.io.File

@Composable
fun InventoryScreen(
    container: AppContainer,
    initialTab: Int = 0,
    onCapture: () -> Unit,
    onOpenItem: (String) -> Unit,
    onOpenListing: (String) -> Unit,
    onOpenBundle: () -> Unit,
    onOpenSettings: () -> Unit,
) {
    val vm: InventoryViewModel = viewModel { InventoryViewModel(container.inventory) }
    val state by vm.state.collectAsStateWithLifecycle()
    val c = snapColors
    var tab by rememberSaveable(initialTab) { mutableIntStateOf(initialTab) }

    Box(Modifier.fillMaxSize().background(c.bg)) {
        Column(Modifier.fillMaxSize()) {
            SnapTopBar(title = "SnapSell", titleStyle = SnapType.appBarTitle.copy(fontSize = 19.sp)) {
                IconBox(R.drawable.ic_lucide_layers, "Build a bundle", onOpenBundle)
                IconBox(R.drawable.ic_lucide_sliders_horizontal, "Settings", onOpenSettings)
            }
            TabRow(
                tabs = listOf("Items" to state.items.size, "Listings" to state.listings.size),
                selected = tab,
                onSelect = { tab = it },
            )
            when (tab) {
                0 -> ItemsList(state.items, onOpenItem)
                else -> ListingsList(state.listings, onOpenListing, onMarkSold = vm::markSold)
            }
        }

        // 64dp square FAB, accent fill, the only shadowed element on this screen.
        Box(
            Modifier
                .align(Alignment.BottomEnd)
                .navigationBarsPadding()
                .padding(end = 16.dp, bottom = 22.dp)
                .shadow(elevation = 12.dp, shape = RectangleShape, ambientColor = c.shadow, spotColor = c.shadow)
                .size(64.dp)
                .background(c.accent)
                .clickable(onClick = onCapture),
            contentAlignment = Alignment.Center,
        ) {
            LucideIcon(R.drawable.ic_lucide_camera, "Capture items", size = 26.dp, tint = c.bg)
        }
    }
}

/** Two flush tabs, 3dp accent underline on the active one, 2dp rule beneath. No indicator animation. */
@Composable
private fun TabRow(tabs: List<Pair<String, Int>>, selected: Int, onSelect: (Int) -> Unit) {
    val c = snapColors
    Column(Modifier.fillMaxWidth()) {
        Row(Modifier.fillMaxWidth()) {
            tabs.forEachIndexed { i, (label, count) ->
                val active = i == selected
                Column(
                    Modifier
                        .weight(1f)
                        .height(50.dp)
                        .clickable { onSelect(i) },
                ) {
                    Box(Modifier.weight(1f).padding(horizontal = 16.dp), contentAlignment = Alignment.CenterStart) {
                        Text(
                            buildAnnotatedString {
                                append(label)
                                withStyle(SnapType.tabLabel.copy(fontWeight = FontWeight.Normal, color = c.text.copy(alpha = 0.55f)).toSpanStyle()) {
                                    append("  $count")
                                }
                            },
                            style = SnapType.tabLabel,
                            color = c.text,
                            modifier = Modifier.alpha(if (active) 1f else 0.5f),
                        )
                    }
                    Box(Modifier.fillMaxWidth().height(3.dp).background(if (active) c.accent else androidx.compose.ui.graphics.Color.Transparent))
                }
            }
        }
        Rule(thickness = 2.dp)
    }
}

@Composable
private fun ItemsList(rows: List<ItemRow>, onOpenItem: (String) -> Unit) {
    if (rows.isEmpty()) {
        EmptyState()
        return
    }
    LazyColumn(Modifier.fillMaxSize(), contentPadding = PaddingValues(bottom = 96.dp)) {
        items(rows, key = { it.item.id }) { row -> ItemRowView(row, onOpenItem) }
    }
}

/** Comfortable row: 14 / 16dp padding, 60dp cutout, 1dp bottom divider. */
@Composable
private fun ItemRowView(row: ItemRow, onOpenItem: (String) -> Unit) {
    val c = snapColors
    val item = row.item
    Column(Modifier.fillMaxWidth().clickable { onOpenItem(item.id) }) {
        Row(Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 14.dp), verticalAlignment = Alignment.CenterVertically) {
            CutoutTile(File(item.photoPath), 60.dp)
            Column(Modifier.weight(1f).padding(horizontal = 12.dp), verticalArrangement = Arrangement.spacedBy(3.dp)) {
                Text(item.name.ifBlank { "Unidentified item" }, style = SnapType.rowTitle, color = c.text, maxLines = 1, overflow = TextOverflow.Ellipsis)
                Text(item.condition.label, style = SnapType.fieldLabel, color = c.text.copy(alpha = 0.6f))
                row.listingStatus?.let { StatusChip(it, Modifier.padding(top = 2.dp)) }
            }
            Column(horizontalAlignment = Alignment.End) {
                val final = item.finalPrice
                val suggested = item.quote?.suggestedPrice
                val shown = final ?: suggested
                Text(if (shown != null) Money.usd(shown) else "—", style = SnapType.rowPrice, color = c.text)
                MicroLabel(
                    when {
                        final != null -> "Final"
                        suggested != null -> "Suggested"
                        item.quote != null -> "No comps"
                        else -> "Not priced"
                    },
                )
            }
        }
        Rule(thickness = 1.dp)
    }
}

@Composable
private fun ListingsList(listings: List<ListingWithItems>, onOpenListing: (String) -> Unit, onMarkSold: (String) -> Unit) {
    if (listings.isEmpty()) {
        EmptyState()
        return
    }
    LazyColumn(Modifier.fillMaxSize(), contentPadding = PaddingValues(bottom = 96.dp)) {
        items(listings, key = { it.listing.id }) { l -> ListingRowView(l, onOpenListing, onMarkSold) }
    }
}

@Composable
private fun ListingRowView(l: ListingWithItems, onOpenListing: (String) -> Unit, onMarkSold: (String) -> Unit) {
    val c = snapColors
    val n = l.items.size
    val meta = buildString {
        append("$n item"); if (n != 1) append('s')
        l.items.firstOrNull()?.let { append(" · ").append(it.condition.label) }
        if (n > 1) append(" · ").append(l.listing.kind.label)
    }
    Column(Modifier.fillMaxWidth().clickable { onOpenListing(l.listing.id) }) {
        Column(Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 14.dp)) {
            Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.Top) {
                Column(Modifier.weight(1f)) {
                    Text(l.listing.title, style = SnapType.rowTitle, color = c.text, maxLines = 2, overflow = TextOverflow.Ellipsis)
                    Text(meta, style = SnapType.fieldLabel, color = c.text.copy(alpha = 0.6f), modifier = Modifier.padding(top = 3.dp))
                }
                Text(Money.usd(l.listing.price), style = SnapType.rowPrice, color = c.text, modifier = Modifier.padding(start = 12.dp))
            }
            Row(Modifier.fillMaxWidth().padding(top = 10.dp), verticalAlignment = Alignment.CenterVertically) {
                StatusChip(l.listing.status)
                if (l.listing.status == ListingStatus.LISTED) {
                    Spacer(Modifier.width(4.dp))
                    GhostButton("Mark as sold", onClick = { onMarkSold(l.listing.id) }, minHeight = 36.dp)
                }
            }
        }
        Rule(thickness = 1.dp)
    }
}

/** Flush left at 64 / 28dp, not centred. */
@Composable
private fun EmptyState() {
    val c = snapColors
    Column(Modifier.fillMaxSize().padding(horizontal = 28.dp, vertical = 64.dp)) {
        Box(Modifier.size(56.dp).border(2.dp, c.text), contentAlignment = Alignment.Center) {
            LucideIcon(R.drawable.ic_lucide_camera, null, size = 26.dp, tint = c.text)
        }
        Text("Nothing here yet", style = SnapType.sectionHead.copy(fontSize = 23.sp, lineHeight = 28.sp), color = c.text, modifier = Modifier.padding(top = 20.dp))
        Text(
            "Point the camera at one item, or at a whole pile. SnapSell separates them out.",
            style = SnapType.bodyLarge,
            color = c.text.copy(alpha = 0.75f),
            modifier = Modifier.padding(top = 8.dp),
        )
    }
}

File: ios/SnapSell/UI/Screens/Inventory/InventoryScreen.swift
import SwiftUI

struct InventoryScreen: View {
    @Environment(AppContainer.self) private var app
    @Environment(Router.self) private var router
    @Environment(\.snap) private var c

    private var statusByItem: [String: ListingStatus] {
        var map: [String: ListingStatus] = [:]
        // Listings are newest-first; keep the first status seen per item.
        for l in app.inventory.listings { for it in l.items where map[it.id] == nil { map[it.id] = l.listing.status } }
        return map
    }

    var body: some View {
        @Bindable var router = router
        ZStack(alignment: .bottomTrailing) {
            VStack(spacing: 0) {
                SnapTopBar("SnapSell", showBack: false, trailing: {
                    iconBox("square.stack.3d.up", "Build a bundle") { router.push(.bundle(itemId: nil)) }
                    iconBox("slider.horizontal.3", "Settings") { router.push(.settings) }
                })
                tabRow
                if router.inventoryTab == 0 { itemsList } else { listingsList }
            }
            Button { router.push(.capture) } label: {
                Image(systemName: "camera.fill").font(.system(size: 26, weight: .bold)).foregroundStyle(c.bg)
                    .frame(width: 64, height: 64).background(c.accent)
                    .shadow(color: .black.opacity(0.25), radius: 12, y: 6)
            }
            .buttonStyle(.plain)
            .accessibilityLabel("Capture items")
            .padding(.trailing, 16).padding(.bottom, 22)
        }
        .screenBackground()
        .toolbar(.hidden, for: .navigationBar)
    }

    private func iconBox(_ symbol: String, _ label: String, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            Image(systemName: symbol).font(.system(size: 20, weight: .semibold)).foregroundStyle(c.text).frame(width: 44, height: 44)
        }
        .buttonStyle(.plain)
        .accessibilityLabel(label)
    }

    private var tabRow: some View {
        VStack(spacing: 0) {
            HStack(spacing: 0) {
                tab(0, "Items", app.inventory.items.count)
                tab(1, "Listings", app.inventory.listings.count)
            }
            Rule()
        }
    }

    private func tab(_ i: Int, _ label: String, _ count: Int) -> some View {
        let active = router.inventoryTab == i
        return Button { router.inventoryTab = i } label: {
            VStack(spacing: 0) {
                HStack(alignment: .firstTextBaseline, spacing: 8) {
                    Text(label).snap(SnapType.tabLabel).foregroundStyle(c.text)
                    Text("\(count)").font(.custom("Archivo-Regular", size: 14)).foregroundStyle(c.text.opacity(0.55))
                    Spacer()
                }
                .padding(.horizontal, 16)
                .frame(maxHeight: .infinity)
                Rectangle().fill(active ? c.accent : .clear).frame(height: 3)
            }
            .frame(height: 50)
        }
        .buttonStyle(.plain)
    }

    private var itemsList: some View {
        let rows = app.inventory.items
        let statuses = statusByItem
        return Group {
            if rows.isEmpty {
                emptyState
            } else {
                ScrollView {
                    LazyVStack(spacing: 0) {
                        ForEach(rows) { item in itemRow(item, status: statuses[item.id]) }
                    }
                    .padding(.bottom, 96)
                }
            }
        }
    }

    private func itemRow(_ item: Item, status: ListingStatus?) -> some View {
        Button { router.push(.item(id: item.id)) } label: {
            VStack(spacing: 0) {
                HStack(spacing: 12) {
                    StoredTile(file: item.photoFile, size: 60)
                    VStack(alignment: .leading, spacing: 3) {
                        Text(item.name.isBlank ? "Unidentified item" : item.name).snap(SnapType.rowTitle).foregroundStyle(c.text).lineLimit(1)
                        Text(item.condition.label).snap(SnapType.fieldLabel).foregroundStyle(c.text.opacity(0.6))
                        if let status { StatusChip(status: status).padding(.top, 2) }
                    }
                    Spacer(minLength: 8)
                    VStack(alignment: .trailing, spacing: 2) {
                        let shown = item.finalPrice ?? item.quote?.suggestedPrice
                        Text(shown.map(Money.usd) ?? "—").snap(SnapType.rowPrice).foregroundStyle(c.text)
                        MicroLabel(
                            item.finalPrice != nil ? "Final" : item.quote?.suggestedPrice != nil ? "Suggested" : item.quote != nil ? "No comps" : "Not priced"
                        )
                    }
                }
                .padding(.horizontal, 16).padding(.vertical, 14)
                Rule(color: c.divider).frame(height: 1)
            }
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
    }

    private var listingsList: some View {
        let listings = app.inventory.listings
        return Group {
            if listings.isEmpty {
                emptyState
            } else {
                ScrollView {
                    LazyVStack(spacing: 0) {
                        ForEach(listings) { l in listingRow(l) }
                    }
                    .padding(.bottom, 96)
                }
            }
        }
    }

    private func listingRow(_ l: ListingWithItems) -> some View {
        let n = l.items.count
        var meta = "\(n) item\(n == 1 ? "" : "s")"
        if let first = l.items.first { meta += " · \(first.condition.label)" }
        if n > 1 { meta += " · \(l.listing.kind.label)" }
        return Button { router.push(.handoff(listingId: l.id)) } label: {
            VStack(spacing: 0) {
                VStack(alignment: .leading, spacing: 10) {
                    HStack(alignment: .top, spacing: 12) {
                        VStack(alignment: .leading, spacing: 3) {
                            Text(l.listing.title).snap(SnapType.rowTitle).foregroundStyle(c.text).lineLimit(2).multilineTextAlignment(.leading)
                            Text(meta).snap(SnapType.fieldLabel).foregroundStyle(c.text.opacity(0.6))
                        }
                        Spacer(minLength: 0)
                        Text(Money.usd(l.listing.price)).snap(SnapType.rowPrice).foregroundStyle(c.text)
                    }
                    HStack(spacing: 12) {
                        StatusChip(status: l.listing.status)
                        if l.listing.status == .listed {
                            Button("Mark as sold") { try? app.inventory.updateListingStatus(l.id, status: .sold) }
                                .snapText(SnapType.buttonLabel).foregroundStyle(c.text).frame(height: 36)
                        }
                        Spacer()
                    }
                }
                .padding(.horizontal, 16).padding(.vertical, 14)
                Rule(color: c.divider).frame(height: 1)
            }
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
    }

    private var emptyState: some View {
        VStack(alignment: .leading, spacing: 0) {
            Image(systemName: "camera").font(.system(size: 26, weight: .semibold)).foregroundStyle(c.text)
                .frame(width: 56, height: 56).overlay(Rectangle().stroke(c.text, lineWidth: SnapMetrics.rule))
            Text("Nothing here yet").font(.custom("Archivo-ExtraBold", size: 23)).foregroundStyle(c.text).padding(.top, 20)
            Text("Point the camera at one item, or at a whole pile. SnapSell separates them out.")
                .snap(SnapType.bodyLarge).foregroundStyle(c.text.opacity(0.75)).padding(.top, 4)
            Spacer()
        }
        .padding(.horizontal, 28).padding(.vertical, 64)
        .frame(maxWidth: .infinity, alignment: .leading)
    }
}

/// CutoutTile backed by the photo store.
struct StoredTile: View {
    @Environment(AppContainer.self) private var app
    var file: String?
    var size: CGFloat = 60
    var selected: Bool = false
    @State private var image: UIImage?
    var body: some View {
        CutoutTile(image: image, size: size, selected: selected)
            .task(id: file) {
                guard let file else { image = nil; return }
                let store = app.photoStore
                image = await Task.detached { store.load(file)?.thumbnail(maxEdge: 300) }.value
            }
    }
}

extension UIImage {
    func thumbnail(maxEdge: CGFloat) -> UIImage {
        let longEdge = max(size.width, size.height)
        guard longEdge > maxEdge else { return self }
        return normalized(maxLongEdge: maxEdge)
    }
}
